src/Mvc/Mvc.Testing/src/WebApplicationFactory.cs

```diff
@@ -192,7 +192,8 @@ private string GetContentRootFromFile(string file)
             var key = typeof(TEntryPoint).Assembly.GetName().FullName;
             try
             {
-                return data[key];
+                // For publish scenarios, ~ should be expanded to the base directory
+                return data[key].Replace("~", AppContext.BaseDirectory);
```

Comment:
Should this perform an exact match (look exactly for `~`)? I would also consider using `TryGetValue`. It's weird that it attempts to catch an exception

```C#
if (!data.TryGetValue(key, out var contentRoot))
{
   throw new KeyNotFoundException($"Could not find content root for project '{key}' in test manifest file '{file}'");
}

if (contentRoot == "~")
{
    contentRoot = AppContext.BaseDirectory;
}
```